### PR TITLE
Added additive attention (linear complexity) to transformer models.

### PR DESCRIPTION
**Pull request by BingzhaoZhu**
*Issue #, if available:*

*Description of changes:*
1. Added additive attention as an option for transformer-based models. To replace the MutiHeadAttention with AdditiveAttention, simply pass the following to TabularPredictor().

```
hyperparameters['FT_TRANSFORMER'] = {
    "model.fusion_transformer.additive_attention": True,
    "model.fusion_transformer.share_qv_weights": False,
}
```
Results using additive attention and its comparison to MultiHeadAttention can be found [here](https://quip-amazon.com/9ixBA110jgY6/Progress-Pre-training-on-Tabular-DL?range_cols=5&range_rows=1#temp:s:temp:C:MVY9817779fa3774e1caae594d75;temp:C:MVY519fd4e2f7184d409214df2af).

2. Since Categorical features are stored as Tuples. Added "tuple" to realtime_predict to fix errors with Categorical features.

3. I believe there is a issue when calculating the n_tokens for CategoricalTransformer. 

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: multimodal/src/autogluon/multimodal/utils/model.py [hunk 1]
```diff
@@ -223,6 +223,8 @@ def create_model(
             embedding_arch=model_config.embedding_arch,
             num_classes=num_classes,
             ffn_d_hidden=OmegaConf.select(model_config, "ffn_d_hidden", default=192),
+            additive_attention=model_config.additive_attention,
+            share_qv_weights=model_config.share_qv_weights,
```
Comment (zhiqiangdon) [inline multimodal/src/autogluon/multimodal/utils/model.py:227]: To be backward compatible, use
 `
additive_attention=OmegaConf.select(model_config, "additive_attention", default=False)
share_qv_weights=OmegaConf.select(model_config, "share_qv_weights", default=False)
`

File: multimodal/src/autogluon/multimodal/utils/model.py [hunk 2]
```diff
@@ -254,6 +256,8 @@ def create_model(
             ffn_d_hidden=OmegaConf.select(model_config, "ffn_d_hidden", default=192),
             num_classes=num_classes,
             cls_token=False,
+            additive_attention=model_config.additive_attention,
+            share_qv_weights=model_config.share_qv_weights,
```
Comment (zhiqiangdon) [inline multimodal/src/autogluon/multimodal/utils/model.py:260]: Ditto.

File: multimodal/src/autogluon/multimodal/utils/model.py [hunk 3]
```diff
@@ -301,6 +305,8 @@ def create_model(
             head_activation=model_config.head_activation,
             adapt_in_features=model_config.adapt_in_features,
             loss_weight=model_config.weight if hasattr(model_config, "weight") else None,
+            additive_attention=model_config.additive_attention,
+            share_qv_weights=model_config.share_qv_weights,
```
Comment (zhiqiangdon) [inline multimodal/src/autogluon/multimodal/utils/model.py:309]: Ditto.

File: multimodal/tests/unittests/test_predictor.py [hunk 4]
```diff
@@ -525,6 +529,8 @@ def test_model_configs():
                 "ffn_activation": "geglu",
                 "head_activation": "relu",
                 "data_types": None,
+                "additive_attention": False,
+                "share_qv_weights": False,
```
Comment (zhiqiangdon) [inline multimodal/tests/unittests/test_predictor.py:533]: No test for `additive_attention ` and `share_qv_weights ` being True.

File: multimodal/src/autogluon/multimodal/models/categorical_transformer.py [hunk 5]
```diff
@@ -108,6 +108,8 @@ def __init__(
         kv_compression_sharing: Optional[str] = None,
         head_activation: Optional[str] = "relu",
         head_normalization: Optional[str] = "layer_norm",
+        additive_attention: Optional[bool] = False,
+        share_qv_weights: Optional[bool] = False,
```
Comment (zhiqiangdon) [inline multimodal/src/autogluon/multimodal/models/categorical_transformer.py:112]: Need docstrings for `additive_attention` and `share_qv_weights`.

File: multimodal/src/autogluon/multimodal/models/ft_transformer.py [hunk 6]
```diff
@@ -396,6 +509,8 @@ def __init__(
         head_normalization: ModuleType,
         d_out: int,
         projection: Optional[bool] = False,
+        additive_attention: Optional[bool] = False,
+        share_qv_weights: Optional[bool] = False,
```
Comment (zhiqiangdon) [inline multimodal/src/autogluon/multimodal/models/ft_transformer.py:513]: Need docstrings for additive_attention and share_qv_weights.

File: multimodal/src/autogluon/multimodal/models/fusion.py [hunk 7]
```diff
@@ -243,6 +243,8 @@ def __init__(
         head_normalization: Optional[str] = "layer_norm",
         adapt_in_features: Optional[str] = None,
         loss_weight: Optional[float] = None,
+        additive_attention: Optional[bool] = False,
+        share_qv_weights: Optional[bool] = False,
```
Comment (zhiqiangdon) [inline multimodal/src/autogluon/multimodal/models/fusion.py:247]: Need docstrings for `additive_attention` and `share_qv_weights`.

File: multimodal/src/autogluon/multimodal/models/numerical_transformer.py [hunk 8]
```diff
@@ -382,6 +382,8 @@ def __init__(
         head_activation: Optional[str] = "relu",
         head_normalization: Optional[str] = "layer_norm",
         embedding_arch: Optional[List[str]] = ["linear"],
+        additive_attention: Optional[bool] = False,
+        share_qv_weights: Optional[bool] = False,
```
Comment (zhiqiangdon) [inline multimodal/src/autogluon/multimodal/models/numerical_transformer.py:386]: Need docstrings for `additive_attention` and `share_qv_weights`.

File: multimodal/tests/unittests/test_predictor.py (hunk 4 above)
Comment (BingzhaoZhu) [reply multimodal/tests/unittests/test_predictor.py:533]: Should I just set "additive_attention" and "share_qv_weights" to be True for fusion transformer? or create new test cases for these two parameters?

File: multimodal/src/autogluon/multimodal/models/ft_transformer.py [hunk 9]
```diff
@@ -311,6 +311,119 @@ def forward(
...
+        ----------
+        d_token:
+            the token size. Must be a multiple of :code:`n_heads`.
+        n_heads:
+            the number of heads. If greater than 1, then the module will have
+            an addition output layer (so called "mixing" layer).
+        bias:
+            if `True`, then input (and output, if presented) layers also have bias.
+            `True` is a reasonable default choice.
+        dropout:
```
Comment (sxjscience) [inline multimodal/src/autogluon/multimodal/models/ft_transformer.py:347]: nit: Need to switch the order of bias and dropout in the docstring.

**Review by sxjscience — APPROVED** [review]
LGTM. I checked the formula of Fastformer and the implementation is correct. Left a small comment about docstring.

File: multimodal/tests/unittests/test_predictor.py (hunk 4 above)
Comment (zhiqiangdon) [reply multimodal/tests/unittests/test_predictor.py]: New test cases would be better. We could remove `clip` in `test_model_config` to accelerate the tests.

File: multimodal/src/autogluon/multimodal/models/categorical_transformer.py [hunk 10]
```diff
@@ -197,7 +203,11 @@ def __init__(
         )
 
         if kv_compression_ratio is not None:
-            n_tokens = self.categorical_feature_tokenizer.n_tokens + 1
+            # number of tokens depends on whether the CLS token is included or not.
+            if self.cls_token:
```
Comment (zhiqiangdon) [inline multimodal/src/autogluon/multimodal/models/categorical_transformer.py:207]: It seems that `if self.cls_token` is always True?

Comment (BingzhaoZhu) [reply multimodal/src/autogluon/multimodal/models/categorical_transformer.py:207]: It is False by default, and should remain False when used together with "fusion_transformer". CLS token is added only once with the fusion transformer.

Comment (zhiqiangdon) [reply multimodal/src/autogluon/multimodal/models/categorical_transformer.py:207]:
```
self.cls_token = (
            CLSToken(
                d_token=d_token,
                initialization=token_initialization,
            )
            if cls_token
            else nn.Identity()
        )
```
According to this, `self.cls_token` seems always True, whether a nn.Identity() or CLSToken().

File: multimodal/src/autogluon/multimodal/models/ft_transformer.py [hunk 11]
```diff
@@ -311,6 +311,119 @@ def forward(
...
+    This implementation is motivated by: https://github.com/jrzaurin/pytorch-widedeep.git
+
+    References:
+    ----------
+    [1] Wu, Chuhan, et al. "Fastformer: Additive attention can be all you need." arXiv preprint arXiv:2108.09084 (2021).
+    """
+
+    def __init__(
+        self,
+        *,
```
Comment (zhiqiangdon) [inline multimodal/src/autogluon/multimodal/models/ft_transformer.py:328]: Why is there a `*` in the arguments?

File: multimodal/src/autogluon/multimodal/models/categorical_transformer.py (hunk 10 above)
Comment (BingzhaoZhu) [reply multimodal/src/autogluon/multimodal/models/categorical_transformer.py:207]: Yes, I think you are right. Should change to cls_token instead.

File: multimodal/src/autogluon/multimodal/models/ft_transformer.py (hunk 11 above)
Comment (BingzhaoZhu) [reply multimodal/src/autogluon/multimodal/models/ft_transformer.py:328]: I noticed MultiHeadAttention did [a similar thing](https://github.com/BingzhaoZhu/autogluon/blob/master/multimodal/src/autogluon/multimodal/models/ft_transformer.py#:~:text=class%20MultiheadAttention(,%2D%3E%20None%3A). Just wish to keep consistent.

File: multimodal/src/autogluon/multimodal/models/ft_transformer.py [hunk 12]
```diff
@@ -311,6 +311,119 @@ def forward(
...
+        for m in trainable:
+            # the "xavier" branch tries to follow torch.nn.MultiheadAttention;
+            # the second condition checks if W_v plays the role of W_out; the latter one
+            # is initialized with Kaiming in torch
+            if initialization == "xavier":
+                # gain is needed since W_qkv is represented with 3 separate layers (it
+                # implies different fan_out)
+                nn.init.xavier_uniform_(m.weight, gain=1 / math.sqrt(2))
+            if m.bias is not None:
+                nn.init.zeros_(m.bias)
```
Comment (zhiqiangdon) [inline multimodal/src/autogluon/multimodal/models/ft_transformer.py:390]: Consider reusing the `init_weights` in `models/utils.py` to initialize the weights. You may need to modify `init_weights` a little bit to do so. You may use `self.apply(init_weights)`.

File: multimodal/src/autogluon/multimodal/models/ft_transformer.py (hunk 11 above)
Comment (zhiqiangdon) [reply multimodal/src/autogluon/multimodal/models/ft_transformer.py:328]: I see. It forces to use keywords when initializing the class.

File: multimodal/src/autogluon/multimodal/utils/inference.py [hunk 13]
```diff
@@ -32,13 +32,13 @@ def extract_from_output(outputs: List[Dict], ret_type: str, as_ndarray: Optional
     """
     if ret_type == LOGITS:
         logits = [ele[LOGITS] for ele in outputs]
-        ret = torch.cat(logits)
+        ret = torch.cat(logits).nan_to_num(nan=-1e4)
```
Comment (zhiqiangdon) [inline multimodal/src/autogluon/multimodal/utils/inference.py:35]: Does nan often appear in the ft transformer outputs? Is the nan in inference normal? Nans in training can cause model collapse.

File: multimodal/src/autogluon/multimodal/models/ft_transformer.py (hunk 12 above)
Comment (BingzhaoZhu) [reply multimodal/src/autogluon/multimodal/models/ft_transformer.py:390]: Should just self.apply(init_weights) be enough? All layers are linear. The gain was set by me not the paper, I guess using the default is also fine.

File: multimodal/src/autogluon/multimodal/utils/inference.py (hunk 13 above)
Comment (BingzhaoZhu) [reply multimodal/src/autogluon/multimodal/utils/inference.py:35]: I chatted with @sxjscience about this. I tested on 104 datasets and the Nan seems to only appear during validation in my experiments.

File: multimodal/src/autogluon/multimodal/models/ft_transformer.py (hunk 12 above)
Comment (zhiqiangdon) [reply multimodal/src/autogluon/multimodal/models/ft_transformer.py:390]: Here you use `initialization == "xavier"` as a condition. Currently, there is not such condition in the `init_weights `. If you still want this condition, you may need to modify it. Otherwise, it's fine to use the default.

File: multimodal/src/autogluon/multimodal/models/ft_transformer.py (hunk 6 above)
Comment (zhiqiangdon) [reply multimodal/src/autogluon/multimodal/models/ft_transformer.py:513]: It would be good if you could add all the missing docstrings. Yiqing forgot doing so in his PR.

Comment (BingzhaoZhu) [reply multimodal/src/autogluon/multimodal/models/ft_transformer.py:502]: This is added. Please check

File: multimodal/src/autogluon/multimodal/models/fusion.py (hunk 7 above)
Comment (BingzhaoZhu) [reply multimodal/src/autogluon/multimodal/models/fusion.py:247]: added

File: multimodal/src/autogluon/multimodal/models/ft_transformer.py (hunk 12 above)
Comment (BingzhaoZhu) [reply multimodal/src/autogluon/multimodal/models/ft_transformer.py]: changed

File: multimodal/src/autogluon/multimodal/models/categorical_transformer.py (hunk 10 above)
Comment (BingzhaoZhu) [reply multimodal/src/autogluon/multimodal/models/categorical_transformer.py]: changed

File: multimodal/tests/unittests/test_predictor.py (hunk 4 above)
Comment (BingzhaoZhu) [reply multimodal/tests/unittests/test_predictor.py]: I think the test case for the new configuration is within ```test_model_config```, since it was added as an argument to ```FT_Transformer```.

**Review by zhiqiangdon — APPROVED** [review]
LGTM! Thanks!
